File: wall_filter/wall_filter_node.py
import rclpy
from rclpy.node import Node

from sensor_msgs.msg import LaserScan, PointCloud2
from std_msgs.msg import String

import numpy as np
import math
import copy
import struct

import tf2_ros as tf2
#from tf2_ros import TransformException
#from tf2_ros.buffer import Buffer
#from tf2_ros.transform_listener import TransformListener

from geometry_msgs.msg import PolygonStamped, Point32, Quaternion#, Rectangle

class WallFilterNode(Node):
    def __init__(self):
        super().__init__('minimal_subscriber')
        # Setup 2D lidar listener / publisher
        self.publisherTransformed = self.create_publisher(PolygonStamped, '/postRotation', 10)
        self.publisherTrimmed = self.create_publisher(LaserScan, '/postRotationLaser', 10)
        self.subscription = self.create_subscription(LaserScan, '/scan', self.laserScan_callback, 10)
        self.subscription

        # Get bounding box
        # Import from kurome/msg in main repo
        #self.arenaSubscriber = self.create_subscription(Rectangle, '/kurome/arena', self.listener_callback, 10)
        self.min_x = -1# + self.r_pose[0]
        self.min_y = -1# + self.r_pose[1]
        self.max_x = 3 #+ self.r_pose[0]
        self.max_y = 0 #+ self.r_pose[1]

        # Setup 3D lidar listener / publisher
        self.pointCloudSubscriber = self.create_subscription(PointCloud2, '/unilidar/cloud', self.point_cloud_callback, 10)
        self.pointCloudPublisher = self.create_publisher(PointCloud2, '/testPointCloud', 100)
        
        # Setup transform listener
        self.tf_buffer = tf2.Buffer()
        self.tf_listener = tf2.TransformListener(self.tf_buffer, self)

        ### Parameters ###
        #self.target_frame = self.declare_parameter('target_frame', 'default_value').get_parameter_value().string_value

        # Should get these from tf listener during callback. Parameters just for testing
        #self.r_pose = [-.5,-.5] 
        #self.quat = [0,0,0,0] # 0 degree rotation
        #self.quat = [0.9659258, 0, 0.258819, 0] # 30 degree rotation

        self.default_bad = 0.0

    def point_cloud_callback(self, message):
        datatype_sizes = {
            1: 1,  # INT8
            2: 1,  # UINT8
            3: 2,  # INT16
            4: 2,  # UINT16
            5: 4,  # INT32
            6: 4,  # UINT32
            7: 4,  # FLOAT32
            8: 8   # FLOAT64
        }

        # Get transform message
        try:
            r_pose = self.tf_buffer.lookup_transform("target_frame", "source_frame", rclpy.time.Time()).transform
            #self.get_logger().info(f'Transform quat: {self.quat}')
        except Exception as ex:
            self.get_logger().info(f'Could not transform : {ex}')
            return

		# Get bounding box relative to lidar device
        box = [ [self.min_x, self.min_y], [self.max_x, self.max_y] ]
        box = self.transform_bounding_box(box, self.quat_to_theta(r_pose.rotation), r_pose.translation)
        
        # Get size of x and y
        x_size = datatype_sizes[message.fields[0].datatype]
        y_size = datatype_sizes[message.fields[1].datatype]
        x = message.fields[0]
        y = message.fields[1]
        
        #valid_points = np.array([]).astype(np.uint8)
        valid_points = []
        
        for i in range(0, message.row_step*message.height, message.point_step):
        	# Create formatter for unpacking ints to float
        	formatter = '>f' if message.is_bigendian else '<f'
        
        	# Get the actual x and y coordinate
        	x_point = struct.unpack(formatter, bytes(message.data[i+x.offset : i+x.offset + x_size]))[0]
        	y_point = struct.unpack(formatter, bytes(message.data[i+y.offset : i+y.offset + y_size]))[0]
        
        	# Check if the point is within bounds
        	if box[0][0] <= x_point <= box[1][0] and box[0][1] <= y_point <= box[1][1]:
        		# Store all data associated with this point
        		valid_points.extend(message.data[i:i+message.point_step])
        
        # Merge new point array into original message and publish
        # Adjust total size of message width
        message.data = valid_points
        message.row_step = int(len(valid_points)) # Might need to force as uint32
        message.width = int(message.row_step / message.point_step) # Might need to force as uint32
        self.pointCloudPublisher.publish(message)

    """
    Triggers when a LaserScan is recieved on /scan
    Trims all ranges outside a bounding box
    """
    def laserScan_callback(self, message):
        #self.get_logger().info("Entire message: {0}\n".format(message))

        # Get transform message
        try:
            r_pose = self.tf_buffer.lookup_transform("target_frame", "source_frame", rclpy.time.Time()).transform
            #self.get_logger().info(f'Transform quat: {self.quat}')
        except Exception as ex:
            self.get_logger().info(f'Could not transform : {ex}')
            return

        # Ensure all points are float32 (required by Polygon) and clean out NaN/infinity values
        points = self.polar_to_cartesian(message.angle_min, message.ranges, message.angle_increment)
        points = np.array(points, dtype=np.float32)
        points_clean = np.nan_to_num(points, nan=self.default_bad, posinf=self.default_bad, neginf=self.default_bad)

        ### TODO remove these two transformations and adjust bounding box instead
        # Apply rotation then shift points according to robot's pose
        transformed = self.apply_rotation(points_clean, self.quat_to_theta(r_pose.rotation))
        transformed = transformed + [r_pose.translation.x, r_pose.translation.y]
        #box = [ [self.min_x, self.min_y], [self.max_x, self.max_y] ]
        #transformed_box = self.transform_bounding_box(box, self.quat_to_theta(r_pose.rotation), r_pose.translation)

        # Cull any points out of bounds
        mask = ((self.min_x+r_pose.translation.x) < transformed[:,0]) & (transformed[:,0] < (self.max_x+r_pose.translation.x)) & ((self.min_y+r_pose.translation.y) < transformed[:,1]) & (transformed[:,1] < (self.max_y+r_pose.translation.y))
        #mask = ( (transformed_box[0][0] < points_clean[:,0]) & (points_clean[:,0] < transformed_box[1][0]) ) & ( (transformed_box[0][1] < points_clean[:,1]) & (points_clean[:,1] < transformed_box[1][1]) )
        culled_points = np.where(mask[:, np.newaxis], transformed, np.array([0,0]))

        # Modify the original laser scan according to the map
        message.ranges = np.where(mask, message.ranges, self.default_bad).astype(np.float32)

        # Publish
        self.publisherTransformed.publish(geo_message)
        self.publisherTrimmed.publish(message)

    """
    Takes in a lidar's LaserScan message and converts it to cartesian coordinates
    """
    def polar_to_cartesian(self, start_angle, distances, angle_step):
        angles = start_angle + np.arange(len(distances)) * angle_step
        x = distances * np.cos(angles)
        y = distances * np.sin(angles)
        return np.column_stack((x, y))
    
    """
    Convert a quaternion to its Yaw angle (theta) so it is usable in a transformation matrix
    """
    def quat_to_theta(self, quat):
        # Assume quaternions are in form here: https://docs.ros.org/en/jade/api/geometry_msgs/html/msg/Quaternion.html
        #t2 = 2.0 * (quat[0] * quat[2] - quat[3] * quat[1])
        t2 = 2.0 * (quat.w * quat.y - quat.z * quat.x)
        t2 = np.where(t2>+1.0,+1.0,t2)
        t2 = np.where(t2<-1.0, -1.0, t2)
        return np.arcsin(t2)

    """
    Multiplies a transformation matrix over all points
    """
    def apply_rotation(self, points, theta):
        tf_matrix = np.array([ [np.cos(theta), -np.sin(theta)], [np.sin(theta), np.cos(theta)] ])
        after_rotation = np.matmul(points,tf_matrix)
        return after_rotation

    """
    Transform the box to the lidar's coordinate frame
    param box should be a 2d array
    """
    def transform_bounding_box(self, box, theta, current_pose):
    	# Corners of the global bounding box
        corners = np.array( [
			[self.min_x, self.min_y],
        	[self.min_x, self.max_y],
        	[self.max_x, self.min_y],
        	[self.max_x, self.max_y]
            ]
		)

    	# Step 1: Translate by -current_pose
        translated_points = corners - np.array([current_pose.x + of, current_pose.y + of])

    	# Step 2: Inverse rotate by -theta
        transformation_matrix = np.array([
            [np.cos(-theta), -np.sin(-theta)],
            [np.sin(-theta), np.cos(-theta)]
            ])
        transformed = np.dot(translated_points, transformation_matrix.T)

		# Return result as [ [min_x, min_y], [max_x, max_y] ]
        new_min = np.min(transformed, axis=0).tolist()
        new_max = np.max(transformed, axis=0).tolist()
        return [ [new_min[0], new_min[1]], [new_max[0], new_max[1]] ]

def main(args=None):
    rclpy.init(args=args)
    wall_filter = WallFilterNode()
    rclpy.spin(wall_filter)
    wall_filter.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/wall_filter_node.cpp
// wall_filter_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/laser_scan.hpp>
#include <sensor_msgs/msg/point_cloud2.hpp>
#include <geometry_msgs/msg/point32.hpp>
#include <geometry_msgs/msg/polygon_stamped.hpp>
#include <kurome/msg/rectangle.hpp>
#include <tf2_ros/transform_listener.h>
#include <tf2_ros/buffer.h>
#include <tf2_sensor_msgs/tf2_sensor_msgs.hpp>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
#include <Eigen/Dense>
#include <cmath>
#include <vector>
#include <algorithm>

using std::placeholders::_1;

class WallFilterNode : public rclcpp::Node {
public:
    WallFilterNode() : Node("wall_filter_node") {
        this->declare_parameter<std::string>("target_frame", "map");
        this->declare_parameter<std::string>("source_frame", "base_link");

        get_parameter("target_frame", target_frame_);
        get_parameter("source_frame", source_frame_);

        default_bad_range_ = 0.0f;
        float default_arena_size = 100000.0f;
        min_x_ = -default_arena_size;
        min_y_ = -default_arena_size;
        max_x_ = default_arena_size;
        max_y_ = default_arena_size;

        scan_sub_ = this->create_subscription<sensor_msgs::msg::LaserScan>("/scan", rclcpp::SensorDataQoS(), std::bind(&WallFilterNode::laser_callback, this, _1));
        kurome_scan_sub_ = this->create_subscription<sensor_msgs::msg::LaserScan>("/kurome/scan", rclcpp::SensorDataQoS(), std::bind(&WallFilterNode::laser_callback, this, _1));
        scan_pub_ = this->create_publisher<sensor_msgs::msg::LaserScan>("/filteredLasers999", 10);

        arena_sub_ = this->create_subscription<kurome::msg::Rectangle>("/kurome/arena", 10, std::bind(&WallFilterNode::arena_callback, this, _1));

        cloud_sub_ = this->create_subscription<sensor_msgs::msg::PointCloud2>("/unilidar/cloud", rclcpp::SensorDataQoS(), std::bind(&WallFilterNode::pointcloud_callback, this, _1));
        cloud_pub_ = this->create_publisher<sensor_msgs::msg::PointCloud2>("/testPointCloud999", 10);

        tf_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
        tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
    }

private:
    void arena_callback(const kurome::msg::Rectangle::SharedPtr msg) {
        min_x_ = msg->arena.origin.x;
        min_y_ = msg->arena.origin.y;
        max_x_ = min_x_ + msg->arena.width;
        max_y_ = min_y_ + msg->arena.height;
    }

    void pointcloud_callback(const sensor_msgs::msg::PointCloud2::SharedPtr msg) {
        geometry_msgs::msg::TransformStamped transform;
        try {
            transform = tf_buffer_->lookupTransform(target_frame_, msg->header.frame_id, tf2::TimePointZero);
        } catch (const tf2::TransformException &ex) {
            RCLCPP_WARN(this->get_logger(), "TF lookup failed: %s", ex.what());
            return;
        }

        sensor_msgs::msg::PointCloud2 transformed_cloud;
        tf2::doTransform(*msg, transformed_cloud, transform);

        std::vector<uint8_t> valid_data;
        valid_data.reserve(transformed_cloud.data.size());

        for (size_t i = 0; i < transformed_cloud.data.size(); i += transformed_cloud.point_step) {
            float x = *reinterpret_cast<const float*>(&transformed_cloud.data[i + transformed_cloud.fields[0].offset]);
            float y = *reinterpret_cast<const float*>(&transformed_cloud.data[i + transformed_cloud.fields[1].offset]);

            if (x >= min_x_ && x <= max_x_ && y >= min_y_ && y <= max_y_) {
                valid_data.insert(valid_data.end(),
                                  transformed_cloud.data.begin() + i,
                                  transformed_cloud.data.begin() + i + transformed_cloud.point_step);
            }
        }

        transformed_cloud.data = valid_data;
        transformed_cloud.row_step = static_cast<uint32_t>(valid_data.size());
        transformed_cloud.width = transformed_cloud.row_step / transformed_cloud.point_step;
        cloud_pub_->publish(transformed_cloud);
    }

    void laser_callback(const sensor_msgs::msg::LaserScan::SharedPtr msg) {
        geometry_msgs::msg::TransformStamped transform;
        try {
            transform = tf_buffer_->lookupTransform(target_frame_, source_frame_, tf2::TimePointZero);
        } catch (const tf2::TransformException &ex) {
            RCLCPP_WARN(this->get_logger(), "TF lookup failed: %s", ex.what());
            return;
        }

        float theta = std::asin(std::clamp(2.0f * (transform.transform.rotation.w * transform.transform.rotation.y -
                                                  transform.transform.rotation.z * transform.transform.rotation.x), -1.0f, 1.0f));

        size_t n = msg->ranges.size();
        std::vector<float> filtered_ranges = msg->ranges;

        float cos_theta = std::cos(theta);
        float sin_theta = std::sin(theta);
        float tx = transform.transform.translation.x;
        float ty = transform.transform.translation.y;

        for (size_t i = 0; i < n; ++i) {
            float r = msg->ranges[i];
            if (!std::isfinite(r)) {
                filtered_ranges[i] = default_bad_range_;
                continue;
            }
            float angle = msg->angle_min + i * msg->angle_increment;
            float x = r * std::cos(angle);
            float y = r * std::sin(angle);

            float x_rot = cos_theta * x - sin_theta * y + tx;
            float y_rot = sin_theta * x + cos_theta * y + ty;

            if (x_rot < min_x_ || x_rot > max_x_ || y_rot < min_y_ || y_rot > max_y_) {
                filtered_ranges[i] = default_bad_range_;
            }
        }

        auto out_scan = *msg;
        out_scan.ranges = filtered_ranges;
        scan_pub_->publish(out_scan);
    }

    std::string target_frame_;
    std::string source_frame_;
    float min_x_, min_y_, max_x_, max_y_;
    float default_bad_range_;

    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr scan_sub_;
    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr kurome_scan_sub_;
    rclcpp::Publisher<sensor_msgs::msg::LaserScan>::SharedPtr scan_pub_;

    rclcpp::Subscription<kurome::msg::Rectangle>::SharedPtr arena_sub_;
    rclcpp::Subscription<sensor_msgs::msg::PointCloud2>::SharedPtr cloud_sub_;
    rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr cloud_pub_;

    std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
    std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
};

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<WallFilterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

